server/src/main/java/com/objectcomputing/checkins/security/ImpersonationController.java

```diff
@@ -0,0 +1,151 @@
+import io.micronaut.http.annotation.Controller;
+import io.micronaut.http.annotation.Get;
+import io.micronaut.http.annotation.Post;
+import io.micronaut.scheduling.TaskExecutors;
+import io.micronaut.scheduling.annotation.ExecuteOn;
+import io.micronaut.security.annotation.Secured;
+import io.micronaut.security.authentication.Authentication;
+import io.micronaut.security.authentication.AuthenticationResponse;
+import io.micronaut.security.authentication.Authenticator;
+import io.micronaut.security.authentication.UsernamePasswordCredentials;
+import io.micronaut.security.event.LoginFailedEvent;
+import io.micronaut.security.event.LoginSuccessfulEvent;
+import io.micronaut.security.handlers.LoginHandler;
+import io.micronaut.security.rules.SecurityRule;
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+
+import java.util.HashMap;
+import java.util.Locale;
+import java.util.Map;
+import java.util.Optional;
+import java.util.HashSet;
+import java.util.Set;
+import java.net.URI;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+@Requires(env = Environments.LOCAL)
+@Controller("/impersonation")
+@ExecuteOn(TaskExecutors.BLOCKING)
+@Secured(SecurityRule.IS_AUTHENTICATED)
+public class ImpersonationController {
+    public static final String JWT = "JWT";
+    public static final String originalJWT = "OJWT";
+    private static final Logger LOG = LoggerFactory.getLogger(ImpersonationController.class);
+    protected final Authenticator authenticator;
+    protected final LoginHandler loginHandler;
+    protected final ApplicationEventPublisher eventPublisher;
+    private final CurrentUserServices currentUserServices;
+    private final SecurityService securityService;
+
+    /**
+     * @param authenticator       {@link Authenticator} collaborator
+     * @param loginHandler        A collaborator which helps to build HTTP response depending on success or failure.
+     * @param eventPublisher      The application event publisher
+     * @param currentUserServices Current User services
+     * @param securityService     The Security Service
+     */
+    public ImpersonationController(Authenticator authenticator,
+                                   LoginHandler loginHandler,
+                                   ApplicationEventPublisher eventPublisher,
+                                   CurrentUserServices currentUserServices,
+                                   SecurityService securityService) {
+        this.authenticator = authenticator;
+        this.loginHandler = loginHandler;
+        this.eventPublisher = eventPublisher;
+        this.currentUserServices = currentUserServices;
+        this.securityService = securityService;
+    }
+
+    @ExecuteOn(TaskExecutors.BLOCKING)
+    @Consumes({MediaType.APPLICATION_FORM_URLENCODED, MediaType.APPLICATION_JSON})
+    @Post("/begin")
+    @RequiredPermission(Permission.CAN_IMPERSONATE_MEMBERS)
+    public Mono<Object> auth(HttpRequest<?> request, String email) {
```

Comment:
You can get rid of the `Mono` usage here. It shouldn't be necessary any longer (there are only a few endpoints that use these still). I think you want an `HttpResponse<Void>` here for your return object instead...or maybe a `HttpResponse<?>`. Then you can use methods like `HttpResponse.serverError()` and `HttpResponse.ok()` to construct a response with the right response code. Those methods return `MutableHttpResponse` instances. So, you can do things like add cookies, etc. if needed.

At minimum it looks like you probably want to return `HttpResponse.serverError()` in the default cases toward the end.

Reply:
The `Authenticator.authenticate()` returns a Publisher.  I'm thinking Mono was left in the LocalLoginController because of this.  Or maybe I just don't know how to deal with the Publisher such that we can return an HttpResponse?

Reply:
I looked at this a little bit. Definitely not worth trying to change.